### PR TITLE
First nVeto monitor plugin

### PR DESCRIPTION
**Pull request by tuankhaibui**
## What does the code in this PR do / what does it improve?
I added a plugin nVetoPeakMonitor at the end of the code.

## Can you briefly describe how it works?
I count number of hitlets_nv per channel and number of events. 

## Can you give a minimal working example (or illustrate with a figure)?
I made an example here https://github.com/XENONnT/analysiscode/tree/master/olmo_nVeto

### CODE REVIEW

File: straxen/plugins/online_monitor.py [hunk 1]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
...
+        'n_nveto_pmts', 
+        type=int,
+        default=120,
+        help='Number of nVeto PMTs'),
+)
+class nVetoPeakMonitor(strax.Plugin):
+    depends_on = 'hitlets_nv', 'events_nv'
+    provides = 'nVeto_peak_monitor'
+    data_kind = 'nVeto_peak_monitor'
+    __version__ = '0.0.1'
```
Comment (jorana) [inline straxen/plugins/online_monitor.py:191]:
```suggestion
    depends_on = ('hitlets_nv', 'events_nv')
    provides = 'online_monitor_nv'
    data_kind = 'online_monitor_nv'
    __version__ = '0.0.2'

```
Maye in order to be consistent with the NVeto naming conventions we could change this better to `online_monitor_nv`

File: straxen/plugins/online_monitor.py [hunk 2]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
...
+            (('End time of the chunk', 'endtime'),
+             np.int64),
+            (('Hitlets per channel', 'hitlets_per_channel'),
+             (np.int64, n_nveto_pmts)),
+            (('Events of nVeto per chunk', 'events_nv_per_chunk'),
+             (np.int64)),
+        ]
+        return dtype
+
+    def compute(self, hitlets_nv, events_nv, start, end):
```
[the remark this answers is not in the document]
Comment (jorana) [reply straxen/plugins/online_monitor.py:208]: This warning I usually ignore as this function is always doing the same for all the plugins

File: straxen/plugins/online_monitor.py [hunk 3]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
...
+        # General setup
+        res = np.zeros(1, dtype=self.dtype)
+        res['time'] = start
+        res['endtime'] = end
+        n_pmt = self.config['n_nveto_pmts']
```
Comment (jorana) [inline straxen/plugins/online_monitor.py:213]:
```suggestion
        min_pmt, max_pmt = self.config['channel_map']['nveto']
        n_pmt = (max_pmt - min_pmt) + 1

```

File: straxen/plugins/online_monitor.py [hunk 4]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
...
         return hist.T
+    
+@export
+@strax.takes_config(
+    strax.Option(
+        'n_nveto_pmts', 
+        type=int,
+        default=120,
+        help='Number of nVeto PMTs'),
+)
```
[the remark this answers is not in the document]
Comment (WenzDaniel) [reply straxen/plugins/online_monitor.py:186]: I think we can remove the option entirely. Like for the TPC the number of nveto pmts can be found in straxen common: https://github.com/XENONnT/straxen/blob/6dcd042b08452eebfb70a9296e65cf76a24c4207/straxen/common.py#L35-L43

Just use 
```
straxen.n_nveto_pmts
```

File: straxen/plugins/online_monitor.py (hunk 3 above)
Comment (WenzDaniel) [reply straxen/plugins/online_monitor.py:213]: See comment above

File: straxen/plugins/online_monitor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (WenzDaniel) [reply straxen/plugins/online_monitor.py:186]: Okay nvm I have not seen that the nveto PMT range is hard coded too. Then we have to keep the option.

File: straxen/plugins/online_monitor.py [hunk 5]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
...
+    def compute(self, hitlets_nv, events_nv, start, end):
+        # General setup
+        res = np.zeros(1, dtype=self.dtype)
+        res['time'] = start
+        res['endtime'] = end
+        n_pmt = self.config['n_nveto_pmts']
+
+        hitlets_channel_count, _ = np.histogram(hitlets_nv['channel'],
+                                                 bins=n_pmt,
+                                                 range=[2000, 2000+n_pmt])
```
[the remark this answers is not in the document]
Comment (WenzDaniel) [reply straxen/plugins/online_monitor.py:217]: Btw something even simpler:

```
channel, counts = np.unique(hitlets_nv['channel'], return_counts=True)
```

**Review by WenzDaniel — APPROVED** [review]
I have here a general question @jorana how does the chunking of the online monitor data work? I always thought that the data is uploaded as soon as a chunk is processed. But does the online monitor time resolution depend on the chunk size? E.g. if the typical chunk size for `events_nv` would allow to store the entire 1h run in a single chunk, does this mean we only have a resolution of 1 h?

**Comment by tuankhaibui** [comment]
@jorana @WenzDaniel : thanks a lot! I am making a new version today. it may take some more time, because I want to cross-check with the notebook that I prepared to plot hitlets and event rates.

File: straxen/plugins/online_monitor.py [hunk 6]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
             range=self.config['area_vs_width_bounds'],
             bins=self.config['area_vs_width_nbins'])
         return hist.T
+
```
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py:178]: I deleted the white space

File: straxen/plugins/online_monitor.py [hunk 7]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
...
+    
+@export
+@strax.takes_config(
+    strax.Option(
+        'n_nveto_pmts', 
+        type=int,
+        default=120,
+        help='Number of nVeto PMTs'),
+)
+class nVetoPeakMonitor(strax.Plugin):
```
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py:187]: I added the docstring at lines 191 - 201 in the updated version

[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py:187]: I changed the class name to OnlineMonitorNV

File: straxen/plugins/online_monitor.py [hunk 8]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
...
+    depends_on = 'hitlets_nv', 'events_nv'
+    provides = 'nVeto_peak_monitor'
+    data_kind = 'nVeto_peak_monitor'
+    __version__ = '0.0.1'
+    rechunk_on_save = False
+
+    def infer_dtype(self):
```
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py:194]: I skip this

File: straxen/plugins/online_monitor.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py:208]: I ignore this

File: straxen/plugins/online_monitor.py [hunk 9]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
...
+
+    def compute(self, hitlets_nv, events_nv, start, end):
+        # General setup
+        res = np.zeros(1, dtype=self.dtype)
+        res['time'] = start
+        res['endtime'] = end
+        n_pmt = self.config['n_nveto_pmts']
+
+        hitlets_channel_count, _ = np.histogram(hitlets_nv['channel'],
+                                                 bins=n_pmt,
```
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py:216]: I ignore this

File: straxen/plugins/online_monitor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py:217]: I ignore this

File: straxen/plugins/online_monitor.py (hunk 1 above)
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py]: I modified at lines 203-206 in the new version.

File: straxen/plugins/online_monitor.py [hunk 10]
```diff
@@ -175,3 +175,47 @@ def area_width_hist(self, data):
             range=self.config['area_vs_width_bounds'],
             bins=self.config['area_vs_width_nbins'])
         return hist.T
+    
+@export
```
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py]: I modified at line 181 in new version

File: straxen/plugins/online_monitor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py]: I modified as @jorana suggested at line 183-189 and import immutabledict at line 3 in new version

File: straxen/plugins/online_monitor.py (hunk 3 above)
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py]: I modified at lines 210-211 and 229-230 in new version

File: straxen/plugins/online_monitor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tuankhaibui) [reply straxen/plugins/online_monitor.py]: I modified as @jorana suggested at lines 233-235 in new version

**Comment by tuankhaibui** [comment]
image
The current online_monitor_nv is ~8kB/8chunks (if I do not misunderstand)

**Review by jorana — APPROVED** [review]
Thanks Khai! Looks good. Would you want to add `online_monitor_nv` in the context where [we start the online monitor](https://github.com/XENONnT/straxen/blob/nVeto_online_monitor/straxen/contexts.py#L195)?

**Comment by jorana** [comment]
> 
> 
> I have here a general question @jorana how does the chunking of the online monitor data work? I always thought that the data is uploaded as soon as a chunk is processed. But does the online monitor time resolution depend on the chunk size? E.g. if the typical chunk size for `events_nv` would allow to store the entire 1h run in a single chunk, does this mean we only have a resolution of 1 h?

Yes it does depend on the live-data chunk size. In fact you can see this when we switch run modes from 5s to 20s chunking. Please keep in mind that the re-chunking happens on saving, not on passing through the data from one plugin to the other. So while we may re-chunk `events_nv` on saving, the `online_monitor_nv` gets data every 5/20 s (depending on the DAQ-settings).

If you re-run it on rechunked data, the result is different.

**Review by jorana — APPROVED** [review]

**Pull request merged**